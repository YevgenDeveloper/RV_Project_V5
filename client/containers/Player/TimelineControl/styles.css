@import 'client/styles/variables';

.timeline {
  position: relative;
  display: block;
  height: 32px;
  width: 100%;
  border-radius: 3px;
  margin: 0 15px;

  &:hover:enabled {
    cursor: pointer;
  }

  &::after {
    content: '';
    position: absolute;
    top: 50%;
    left: 0;
    width: 100%;
    height: 4px;
    margin-top: -2px;
    border-radius: 2px;
    background-color: rgba(255, 255, 255, 0.3);
  }
}

.path {
  z-index: 1;
  position: absolute;
  top: 50%;
  left: 0;
  width: 0;
  height: 4px;
  margin-top: -2px;
  border-radius: 2px 0 0 2px;
  background-color: $color-blue;
}

.slider {
  z-index: 2;
  position: absolute;
  top: 4px;
  width: 4px;
  height: 24px;
  margin-left: -2px;
  border-radius: 1px;
  background-color: white;
}
