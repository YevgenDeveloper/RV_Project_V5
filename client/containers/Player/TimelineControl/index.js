// @flow

import React, { Component } from 'react';

import styles from './styles.css';

type Props = {
  currentImage: number,
  totalImages: number,
  disabled: boolean,
  onChangeImage: number => void,
};

type State = {
  totalImages: number,
  pixelsPerFrame: number,
  currentImage: number,
  sliderPosition: number,
};

export default class TimelineControl extends Component<Props, State> {
  state = {
    totalImages: 0,
    pixelsPerFrame: 1,
    // eslint-disable-next-line
    currentImage: 0,
    sliderPosition: 0,
  };

  componentDidMount() {
    document.addEventListener('resize', this.updateSlider);
    this.updateSlider();
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (
      prevProps.currentImage !== this.props.currentImage ||
      prevState.totalImages !== this.state.totalImages
    ) {
      this.updateSlider();
    }
  }

  componentWillUnmount() {
    document.removeEventListener('resize', this.updateSlider);
  }

  static getDerivedStateFromProps(nextProps: Props) {
    return {
      currentImage: nextProps.currentImage,
      totalImages: nextProps.totalImages,
    };
  }

  updateSlider = () => {
    this.setState(state => {
      if (!state.totalImages) {
        return null;
      }

      const pixelsPerFrame = this.ref.offsetWidth / state.totalImages;

      return {
        pixelsPerFrame,
        sliderPosition: state.currentImage * pixelsPerFrame,
      };
    });
  };

  handleClick = (event: MouseEvent) => {
    const { left } = this.ref.getBoundingClientRect();
    const sliderPosition = event.pageX - left;

    this.setState({ sliderPosition }, () => {
      this.props.onChangeImage(
        Math.round(sliderPosition / this.state.pixelsPerFrame),
      );
    });
  };

  ref: HTMLButtonElement;

  render() {
    const { sliderPosition } = this.state;

    return (
      <button
        type="button"
        ref={ref => {
          if (ref) this.ref = ref;
        }}
        className={styles.timeline}
        onClick={this.handleClick}
        disabled={this.props.disabled}
      >
        <div
          className={styles.path}
          style={{ width: `${sliderPosition > 4 ? sliderPosition - 4 : 0}px` }}
        />
        <div
          className={styles.slider}
          style={{ transform: `translateX(${sliderPosition}px)` }}
        />
      </button>
    );
  }
}
