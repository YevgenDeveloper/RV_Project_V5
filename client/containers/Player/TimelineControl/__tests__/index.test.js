// @flow

import type { ShallowWrapper } from 'enzyme';

import React from 'react';
import { shallow } from 'enzyme';
import TimelineControl from '..';

it('renders correctly', () => {
  const wrapper: ShallowWrapper = shallow(
    <TimelineControl disabled={false} onChangeImage={jest.fn()} />,
  );

  expect(wrapper).toMatchSnapshot();
});
