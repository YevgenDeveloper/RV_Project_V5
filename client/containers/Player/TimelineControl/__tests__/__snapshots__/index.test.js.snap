// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimelineControl
    disabled={false}
    onChangeImage={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="path"
          style={
            Object {
              "width": "0px",
            }
          }
        />,
        <div
          className="slider"
          style={
            Object {
              "transform": "translateX(0px)",
            }
          }
        />,
      ],
      "className": "timeline",
      "disabled": false,
      "onClick": [Function],
      "type": "button",
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "path",
          "style": Object {
            "width": "0px",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "slider",
          "style": Object {
            "transform": "translateX(0px)",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="path"
            style={
              Object {
                "width": "0px",
              }
            }
          />,
          <div
            className="slider"
            style={
              Object {
                "transform": "translateX(0px)",
              }
            }
          />,
        ],
        "className": "timeline",
        "disabled": false,
        "onClick": [Function],
        "type": "button",
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "path",
            "style": Object {
              "width": "0px",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "slider",
            "style": Object {
              "transform": "translateX(0px)",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
